lib/record_store/zone.rb

```diff
@@ -272,6 +273,26 @@ def validate_provider_can_handle_zone_records
       end
     end
 
+    def validate_zone_record_not_shadowed
+      nameserver_records = records
+        .select { |record| record.is_a?(Record::NS) && name != record.fqdn }
+        .map { |record| record.fqdn.delete_suffix(".") }
+        .uniq
+
+      nameserver_records.each do |ns_record|
+        selected_records = records.reject do |record|
+          record.is_a?(Record::NS) && \
+            record.fqdn.delete_suffix(".") == ns_record
+        end
+        selected_records.each do |record|
+          normalized_record = record.fqdn.delete_suffix(".")
+          next unless normalized_record.end_with?(".#{ns_record}") || normalized_record == ns_record
```

Comment:
Could even slot a `.filter { ... }` into this collection pipeline perhaps, instead of the `next`